compiler/plugins/target/AMD-AIE/iree-amd-aie/Target/XCLBinGen.cpp

```diff
@@ -247,6 +247,46 @@ FailureOr<std::vector<std::string>> makePeanoOptArgs(
       *iter = flag;
     }
   }
+
+  // Some background to this opt pass finetuning/hack.
+  // currently performance (function outlining, phoenix, matmul) is much better
+  // with -O3 than -O2. But the memory used with -O3 is much higher than with
+  // -O2. This is because with -O3, the outlined function is inlined at most
+  // call sites and in the the process of inlining for some reason the integer
+  // arithmetic is rearranged, and in particular there are fewer common
+  // sub-expressions. For example in the -O2 case, the outlined function after
+  // optimization contains
+  //
+  //    ```
+  //    ...
+  //    %45 = trunc i64 %38 to i20
+  //    ...
+  //    %49 = trunc i64 %38 to i20
+  //    ...
+  //    ```
+  //
+  // whereas the inlined (with -O3) does not. So inlining has found a roundabout
+  // way to improve the code which has nothing to do with any actual traditional
+  // advantages of inlining (fewer function calls etc).
+  //
+  // Can we get this improvement without the memory overhead of -O3? I have
+  // found that running the -O2 pass pipeline, followed by a basic llvm cse pass
```

Comment:
I am likely missing something, but where exactly are we changing `O3` to `O2`? I still see `O3` here: https://github.com/nod-ai/iree-amd-aie/blob/d79bb60b662d8f1d79e5bbd63b673b9e401eee13/compiler/plugins/target/AMD-AIE/iree-amd-aie/Target/XCLBinGen.cpp#L215

Reply:
O3 is still the default, I haven't changed that in this PR. This PR just makes O2 better. There are tests in run.py which use O2 by explicitly opting in for it. 

I could make O2 the default in this PR, but I'd rather do it in a separate PR.

Reply:
Ok, I see, I thought that the point of this PR was to make a case to switch to O2.